### PR TITLE
Add support for MintMe.com Coin

### PR DESCRIPTION
**Pull request by lukaszmatczak**
This pull request adds support for MintMe.com Coin and its lyra2 consensus: https://www.mintme.com/coin/
core-geth will support new fresh blockchain after reset (date and block number not known yet) and preserve current accounts.

Work is in progress. Any comments are welcome.

### CODE REVIEW

**Comment by meowsbits** [comment]
This is looking very strong... Nice work! :+1:

I'll continue to review this in more detail, and will keep an eye on your progress. Please feel free to ping me if you have any questions.

**Comment by meowsbits** [comment]
Just an FYI, at this moment #357 is our latest "staging" branch; it includes merged changes from ethereum/go-ethereum through 1.10.2, which include Berlin protocol features. We're not rushing (yet) to bring this to master because of some concerns about inherited upstream stability issues, but we're putting it through its paces in development now. 

Given a 10 minute glance at your patch, I don't foresee any major conflicts or headaches (since you've namespaced everything quite nicely) so I wouldn't worry about it yet if Berlin features aren't an immediate concern for MINTME.

**Comment by meowsbits** [comment]
@lukaszmatczak Is this still a WIP?

**Comment by lukaszmatczak** [comment]
@meowsbits Yes. We scheduled stopping old blockchain on 2021-06-14. Then I'll update prealloc and this PR will be ready to merge.

**Comment by lukaszmatczak** [comment]
Prealloc and genesis block are updated now. This PR is ready to merge.

**Comment by lukaszmatczak** [comment]
@meowsbits Is everything OK with this PR? If so, when can we expect it to be merged?

**Review by meowsbits — CHANGES_REQUESTED** [review]
- [x] Using the `console`, the module `lyra2` is unavailable. Have not checked other RPC transports.

```
./build/bin/geth attach /tmp/geth.ipc
Welcome to the Geth JavaScript console!

instance: CoreGeth/v1.12.1-unstable-9d36627a-20210615/linux-amd64/go1.16.2
at block: 0 (Mon Jun 14 2021 04:57:53 GMT-0500 (CDT))
 datadir: /home/ia/sandbox/etclabscore/core-geth/datadir
 modules: admin:1.0 debug:1.0 eth:1.0 lyra2:1.0 miner:1.0 net:1.0 personal:1.0 rpc:1.0 trace:1.0 txpool:1.0 web3:1.0

To exit, press ctrl-d
> lyra2
ReferenceError: lyra2 is not defined
        at <eval>:1:1(0)

> debug
{
  accountRange: function(),
  backtraceAt: function(),
```

- [x] The `cmd/devp2p/nodesetcmd.go` should be updated to include a switch statement for `MintMe` for nodeset filtering.
- [x] Currently `forkid` tests (see `core/forkid_test.go`) are not implemented for MintMe. Is MintMe expected to use `eth/65` and up? 
- [x] Is mining supported (and/or supposed to be)? Can you please help educate us about what to expect (if and) when mining with Lyra2 on MintMe? Does it have a DAG? Does the footprint change at all? Are any existing `.ethash` stores preserved unchanged? 
- [x] Can you please link to the original source for the Lyra2 C implementation? We need to have a citation at least to confirm license compatibility.

**Comment by meowsbits** [comment]
This is really looking good, @lukaszmatczak.

Before merging, we need to establish expectations clearly. 
Once merged, this feature set will be your responsibility. If MintMe upgrades or changes in any way, maintenance of the corresponding code will be left to you. If upstream or local changes cause need for significant changes to the MintMe facets of the code, we'll expect you to be available and willing to help with that.  If the feature is abandoned, disused, or otherwise ineffectual, we need to reserve the right to remove the feature without warming. 

I hope this is no surprise to you, and only basically boilerplate contributor guidelines. 

Thanks again for your involvement; I'm excited to see this merged.

**Comment by lukaszmatczak** [comment]
> * [ ]  Using the `console`, the module `lyra2` is unavailable. Have not checked other RPC transports.

I've removed lyra2 namespace at all. It's not needed I think.

> * [ ]  The `cmd/devp2p/nodesetcmd.go` should be updated to include a switch statement for `MintMe` for nodeset filtering.

Done

> * [ ]  Currently `forkid` tests (see `core/forkid_test.go`) are not implemented for MintMe.

Done. 

> Is MintMe expected to use `eth/65` and up?

Yes.

> * [ ]  Is mining supported (and/or supposed to be)? Can you please help educate us about what to expect (if and) when mining with Lyra2 on MintMe? Does it have a DAG? Does the footprint change at all? Are any existing `.ethash` stores preserved unchanged?

Yes, mining is supported. lyra2 is doesn't require any additional dependencies like DAG or persistent storage. Only CPU and RAM. It shouldn't touch any ethash data.

> * [ ]  Can you please link to the original source for the Lyra2 C implementation? We need to have a citation at least to confirm license compatibility.

https://github.com/leocalm/Lyra/tree/master/Lyra2/src

**Comment by ziogaschr** [comment]
Nice work @lukaszmatczak. I reviewed the code as well and it looks good. I was able to sync, start mining, although no lucky to mine a block.

On top of @meowsbits remarks, I will only add that all caps for **MINTME** vars in code looks a bit too much maybe. For example `MINTMEGenesisHash`. I am wondering if we can change this?

**Comment by meowsbits** [comment]
Good idea, too -- `MINTME` can become `MintMe` (using CamelCase, like the rest).

**Comment by lukaszmatczak** [comment]
@ziogaschr @meowsbits Coin name is `MintMe.com Coin` actually (to distinguish from `MintMe.com` exchange), so I've decided to use its ticker - `MINTME`. That's why it's uppercase where it was possible. But if you still think that it should be changed, please let me know, I'll do it.

**Comment by ziogaschr** [comment]
@lukaszmatczak thanks for explaining and reasoning about. It seems to me that in both cases we are talking about the same domain for both coin and exchange. As well they are both under the umbrella of the same company/chain/etc. Based on that and if @meowsbits also agree I will prefer it to be CamelCase in the scope of core-geth. It's totally fine to keep it like `MINTME` on our docs or in command flag help text if you still wish.

Thanks for your understanding.

**Comment by lukaszmatczak** [comment]
@ziogaschr OK. I changed it.

File: core/forkid/forkid_test.go [hunk 1]
```diff
@@ -223,6 +223,15 @@ func TestCreation(t *testing.T) {
 				{3_985_894, ID{Hash: checksumToBytes(0x92b323e0), Next: 0}},
 			},
 		},
+		// MintME test cases
```
Comment (ziogaschr) [inline core/forkid/forkid_test.go]:
```suggestion
		// MintMe test cases
```

**Review by ziogaschr — APPROVED** [review]
LGTM. Nice work @lukaszmatczak

**Comment by meowsbits** [comment]
On the merge, let's make it a squash since the commit messages aren't conventional and the change set as a whole makes sense.

**Review by meowsbits — APPROVED** [review]

**Pull request merged**